### PR TITLE
image preprocessing bindings and tests

### PR DESCRIPTION
**Pull request by segalmax**

### CODE REVIEW

File: py_bind/optv/image_processing.pyx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+from optv.parameters cimport ControlParams, control_par
+import numpy as np
+from twisted.cred import error
```
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: What's this ``twisted`` import?

File: py_bind/optv/image_processing.pyx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+cimport numpy as np
+
+def preprocess_image(np.ndarray[ndim=2, dtype=np.uint8_t] input,
+                   int filter_hp,
+                   ControlParams control,
+                   int lowpass_dim=0,
```
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: Didn't we find that this default is an error, that it should be at least 1?
I think lowpass_dim=0 would make this an identity function.

File: py_bind/optv/image_processing.pyx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    particle detection: an averaging (smoothing) filter on an image, optionally
+    followed by additional user-defined filter.
+    
+    Arguments:
+    numpy.ndarray input - numpy 2d array representing the source image to filter.
+    int filter_hp - flag for additional filtering of _hp. 1 for lowpass, 2 for 
+        general 3x3 filter given in parameter ``filter_file``.
+    ControlParams control - image details such as size and image half for 
+    interlaced cases.
+    int lowpass_dim - dimension of subtracted lowpass image.
```
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: No it isn't. It's the half-width of the box_blur filter.

File: py_bind/optv/image_processing.pyx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    numpy.ndarray output - result numpy 2d array representing the source 
+        image to filter. Same size as img.
+    
+    Returns:
+    numpy.ndarray representing the result image.
+    '''
+
+    # check arrays dimensions
+    if input.ndim != 2:
+        raise TypeError("Input matrix must be two-dimensional")
```
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: A 2D array is different than a matrix both in the class used, the handling of dimensions and in the default methods for multiplication etc. Please keep that in mind. And here it should say 'array'.

File: py_bind/test/test_image_processing.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        res = preprocess_image(self.input,
+                               self.filter_hp,
+                               self.control,
+                               lowpass_dim=1,
+                               filter_file=None,
+                               output=None)
+        
+        np.testing.assert_array_equal(res, correct_res)
+
+    def tearDown(self):
```
Comment (yosefm) [inline py_bind/test/test_image_processing.py]: If it's empty, delete it.

**Comment by yosefm** [comment]
All the tests are missing, I think, is more documentation of why you do what you do. Foe one, it should be mentioned somewhere that some values are taken directly from the C tests.

**Comment by yosefm** [comment]
The PR generally works, so just address the comments and we'll get it upstream.

The merge conflicts come from changes to the setup file that were introduced recently, so for other tasks make sure you work from most recent master. For this one I can handle the conflict.

File: py_bind/optv/image_processing.pyx (hunk 3 above)
Comment (segalmax) [inline py_bind/optv/image_processing.pyx]: This is the value passed as dim_lp in 
```
int prepare_image(unsigned char * img,
                        unsigned char * img_hp,
                        int dim_lp,
                        int filter_hp,
                        char * filter_file,
                        control_par * cpar)
```
and the C documentation states: "int dim_lp - dimension of subtracted lowpass image.".
So where is the mistake?

File: py_bind/optv/image_processing.pyx [hunk 6]
```diff
@@ -24,38 +23,38 @@ def preprocess_image(np.ndarray[ndim=2, dtype=np.uint8_t] input,
...
     Returns:
     numpy.ndarray representing the result image.
     '''
 
     # check arrays dimensions
-    if input.ndim != 2:
-        raise TypeError("Input matrix must be two-dimensional")
-    if output != None and (input.shape[0] != output.shape[0] or
-                         input.shape[1] != output.shape[1]):
+    if input_img.ndim != 2:
```
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: Should be 'is None', not != None. We get a future warning now, that you should have seen when running the tests. This is something I already fixed in other bindings you wrote, and I'll fix it now before submitting upstream, but please pay attention to warnings.

File: py_bind/optv/image_processing.pyx (hunk 3 above)
Comment (yosefm) [inline py_bind/optv/image_processing.pyx]: The mistake is in the C documentation, as I told you when we discussed it in my office - remember the part about how filters work?
Anyway, I fixed it in the upstream PR, just FYI.

**Pull request merged**